src/blr/cli/clusterrmdup.py

```diff
@@ -12,237 +12,297 @@
 import pysam
 import logging
 from tqdm import tqdm
-from collections import Counter
+from collections import Counter, deque, OrderedDict
 
 from blr import utils
 
 logger = logging.getLogger(__name__)
-summary = Counter()
+
+BUFFER_SIZE = 500
 
 
 def main(args):
     logger.info("Starting Analysis")
-
-    current_cache_rp = dict()
-    cache_reads = dict()
+    summary = Counter()
+    positions = OrderedDict()
     chrom_prev = None
-    pos_prev = None
+    pos_prev = 0
     merge_dict = dict()
-    cache_dup_pos = dict()
-    with pysam.AlignmentFile(args.input, "rb") as openin:
-        for read in tqdm(openin.fetch(until_eof=True), desc="Processing reads"):
-            summary["Total reads"] += 1
-
-            # Wait for mate of read until process
-            if read.query_name not in cache_reads:
-                cache_reads[read.query_name] = read
-                continue
-            else:
-                mate = cache_reads[read.query_name]
-                del cache_reads[read.query_name]
+    buffer_dup_pos = deque()
+    for read, mate in tqdm(parse_filtered_read_pairs(args.input, summary), desc="Reading pairs"):
```

Comment:
I think this makes it seem as the function expects a filtered input, how about renaming it as `parse_and_filter_read_pairs`?

Reply:
True